File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.py
import sys
input = sys.stdin.readline

n = int(input())
arr = [] # [fruit, cnt]
idx = -1
for f in map(int, input().split()):
    if not arr or arr[idx][0] != f:
        arr.append([f, 1])
        idx += 1
    else: arr[idx][1] +=1

s, e, ans = 0, 0, 0
while s < len(arr):
    f1 = arr[s][0]
    cnt = arr[s][1]
    m, f2 = -1, -1
    for e in range(s+1, len(arr)):
        if arr[e][0] != f1 and m == -1:
            m = e
            f2 = arr[e][0]
        if arr[e][0] == f1 or arr[e][0] == f2:
            cnt += arr[e][1]
            if e-m > 0:m+=1
        else:break
    ans = max(ans, cnt)
    if s != m and m != -1:s = m
    else: s += 1

print(ans)

File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.java
import java.util.StringTokenizer;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine()), answer = 0, f1 = 0, f2 = 0, cnt = 0, idx = 0;
		int[] arr = new int[n];
		
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");
		for (int i = 0; i < n; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
			if (f1 == 0) f1 = arr[i];
			else if (f1 != arr[i] && f2 == 0) f2 = arr[i];
			else if (f1 != arr[i] && f2 != arr[i]) {
				answer = Math.max(answer, cnt);
				cnt = i - idx;
				f1 = arr[idx];
				f2 = arr[i];
			}
			
			if (arr[i] == f1 || arr[i] == f2) {
				cnt++;
				if (i > 0 && arr[i-1] != arr[i]) idx = i;
			}
		}
		answer = Math.max(answer, cnt);
		System.out.println(answer);
		br.close();
	}
}